### PR TITLE
[tampermonkey] Add missing properties and new methods, update jsdoc comments

### PR DESCRIPTION
**Pull request by double-beep**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [TM docs](https://www.tampermonkey.net/documentation.php)
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

File: types/tampermonkey/index.d.ts [hunk 1]
```diff
@@ -376,21 +420,149 @@ declare namespace Tampermonkey {
...
+             */
+            redirect?: string | {
+                /** A RegExp to extract some parts of the URL (for example `"([^:]+)://match.me/(.*)"`). */
+                from: string;
+                /** Pattern for substitution (for example `"$1://redirected.to/$2"`). */
+                to: string;
+            };
+        };
+    }
+    type WebRequestListener = (
```
Comment (nikolay-borzov) [inline types/tampermonkey/index.d.ts:459]: Please add an empty line between interface/type declarations

File: types/tampermonkey/index.d.ts [hunk 2]
```diff
@@ -376,21 +420,149 @@ declare namespace Tampermonkey {
         version?: string;
     }
 
-    type ContentType = string | { type?: string | undefined; mimetype?: string | undefined };
+    type ContentType = string | { type?: string; mimetype?: string };
+
+    // GM_webRequest
```
Comment (nikolay-borzov) [inline types/tampermonkey/index.d.ts:425]: Should it be JSDoc comment? Can be useful

File: types/tampermonkey/index.d.ts [hunk 3]
```diff
@@ -432,99 +603,237 @@ interface Window {
...
+ * @url https://www.tampermonkey.net/documentation.php#api:GM_getValue
+ * @param name A string specifying the key for which the value should be retrieved.
+ * @param defaultValue A default value to be returned if the key does not exist in the extension's storage.
+ * @returns The value of the specified key from the extension's storage, or the default value if the key does not exist.
+ */
 declare function GM_getValue<TValue>(name: string, defaultValue?: TValue): TValue;
 
-/** Deletes 'name' from storage */
+/**
+ * Deletes `name` from the userscript's storage.
```
Comment (nikolay-borzov) [inline types/tampermonkey/index.d.ts:696]: Missing `@param name`?

File: types/tampermonkey/index.d.ts [hunk 4]
```diff
@@ -432,99 +603,237 @@ interface Window {
...
+/**
+ * Saves information about a tab for later use.
+ * @url https://www.tampermonkey.net/documentation.php#api:GM_saveTab
+ * @param tab An object containing the information to be saved about the tab.
+ */
+declare function GM_saveTab(tab: object): void;
 
-/** Gets a object that is persistent as long as this tab is open */
+/**
+ * Gets a object that is persistent as long as this tab is open
```
Comment (nikolay-borzov) [inline types/tampermonkey/index.d.ts:814]: Missing `@param obj`?

File: types/tampermonkey/index.d.ts [hunk 5]
```diff
@@ -432,99 +603,237 @@ interface Window {
...
+ * @url https://www.tampermonkey.net/documentation.php#api:GM_getTab
+ */
 declare function GM_getTab(callback: (obj: any) => void): void;
 
-/** Gets all tab objects as a hash to communicate with other script instances */
+/**
+ * Gets all tab objects as a hash to communicate with other script instances
```
Comment (nikolay-borzov) [inline types/tampermonkey/index.d.ts:820]: Missing `@param callback`?

File: types/tampermonkey/index.d.ts [hunk 6]
```diff
@@ -535,30 +844,44 @@ declare function GM_log(...message: any[]): void;
...
- * @param text Text of the notification
- * @param title Notification title. If not specified the script name is used
- * @param onclick Called in case the user clicks the notification
+ * Shows an HTML5 Desktop notification and/or highlight the current tab
+ * using a provided message and other optional parameters.
+ * @url https://www.tampermonkey.net/documentation.php#api:GM_notification
+ * @param text A string containing the message to display in the notification.
+ * @param title The title of the notification. If not specified, the script name is used.
+ * @param image The URL of an image to display in the notification.
+ * @param onclick A callback function that will be called when the user clicks on the notification.
```
Comment (nikolay-borzov) [inline types/tampermonkey/index.d.ts:884]: Could we use camelCase here - `onClick`?

**Review by nikolay-borzov — COMMENTED** [review]
Good work! A few comments

File: types/tampermonkey/index.d.ts (hunk 2 above)
Comment (double-beep) [reply types/tampermonkey/index.d.ts:425]: I'm not sure what should be commented. Are you referring to the `ContentType` type?

Comment (nikolay-borzov) [reply types/tampermonkey/index.d.ts:425]: I meant something like this:
```
/**
* GM_webRequest
*/
interface WebRequestRuleParam
```
So user can know what `WebRequestRuleParam` is used for when importing type directly.
Feel free to skip if it makes little sense to you

**Review by nikolay-borzov — APPROVED** [review]

**Comment by double-beep** [comment]
Ready to merge

**Pull request merged**